includes/internal/emails/class-email-generator.php

```diff
@@ -44,16 +44,27 @@ public function student_started_course_mail_to_teacher( $student_id, $course_id
 			return;
 		}
 
-		$teacher   = new \WP_User( $course->post_author );
-		$student   = new \WP_User( $student_id );
-		$recipient = stripslashes( $teacher->user_email );
+		$teacher    = new \WP_User( $course->post_author );
+		$student    = new \WP_User( $student_id );
+		$recipient  = stripslashes( $teacher->user_email );
+		$manage_url = esc_url(
```

Comment:
Looking at these changes, I think our current solution might be not flexible enough.
Maybe we can delegate preparing values to emails (or whatever) by adding a hook here.

Reply:
The problem with current solution, as I see it, is that we have to prepare all the possible placeholders.
But our emails know what exactly they need.
We have an option to add a lot of conditions depending on email name or delegate this work to emails themselves (or any other delegats). Here is a rough draft how it could look like.

```php
class Email_Generator {
	public function student_started_course_mail_to_teacher( $student_id, $course_id ) {
		// ...
		$replacements = apply_filter( 'sensei_emails_placeholders', [], $email_name, $student_id, $course_id );
		// ...
	}
}

class Student_Starts_Course_Email {
	public function init() {
		add_filter( 'sensei_email_placeholders', [ $this, 'prepare_placeholders' ], 10, 4 );
	}

	public function prepare_placeholders( $replacements, $email_name, $student_id, $teacher_id ) {
		$teacher    = new \WP_User( $course->post_author );
		$student    = new \WP_User( $student_id );
		$recipient  = stripslashes( $teacher->user_email );
		$manage_url = esc_url(
			add_query_arg(
				array(
					'page'      => 'sensei_learners',
					'course_id' => $course_id,
					'view'      => 'learners',
				),
				admin_url( 'admin.php' )
			)
		);

		$replacements['student:displayname'] = $student->display_name;
		$replacements['course:name']         = $course->post_title;
		$replacements['manage:students']     = $manage_url;

		return $replacements;
	}
}
```

Reply:
Just noticed that this method isn't generic, but for a specific email.
Anyway, the approach described above could be used to implement a generic one without duplication :)

Reply:
> But our emails know what exactly they need.
@merkushin . For now, is expected to have the user any placeholder data except if we find any strong reason to not do it.